# 前言

我们在学习过程中经常会看到这样或者那样的错误，学会读懂错误，也是很重要的。

最常见的几种错误：

- 语法错误
- 名字错误
- 类型错误
- 文件不存在错误

在编程过程中，出现错误是难免的，尤其是你在运行一些别人的代码的时候。这节课，我们就来讲讲 Python 的错误和异常该如何处理。

本节课程你将学到：



In [4]:
print("hello world

SyntaxError: EOL while scanning string literal (<ipython-input-4-cc6d3073532a>, line 1)

# 解读

观察上面的代码和输出，最后一行最明显，Python 程序提示出现了一个 SyntaxError，也就是「语法错误」。

同时你看倒数第二行，可以看到有个小箭头，指向我们错误发生的位置，在world后面。

我们继续看最后一行关于错误的具体描述，「在扫描字符串的时候遇到了 EOL」。这里的 EOL 是 end of line（行尾）的缩写。

而第一行，是说明文件名和行号，错误发生在哪个文件的哪行代码。

## 语法错误（SyntaxError）

语法错误一般比较容易发现。

## 代码异常（Exception）

而更多的时候，是语法没有问题，但是运行起来出现错误。

我们来看几个异常的错误，这些都是来自官方的文档。

In [6]:
# 案例1：把 0 当作除数
10*(1/0)

ZeroDivisionError: division by zero

In [7]:
# 案例2：spam 名字尚未定义
4+spam+3

NameError: name 'spam' is not defined

In [10]:
# 案例3：类型不同
'2'+2

TypeError: can only concatenate str (not "int") to str

# 解读

以上三段代码一共出现三种异常，分别是：

1. ZeroDivisionError（除零异常）
2. NameError（名字异常）
3. TypeError（类型异常）

以上都是异常的类型，其实这里涉及到了面向对象编程里面的类，这个在 Python 进阶里面的[类]()会讲到。

我们还是回过来看看上面每个异常的具体描述，三个分别是什么。

同样的，我们也看到除了出错的位置，还有一个叫「Traaceback」的提示，它的意思就是追溯（最近一次的执行），而且阅读的方式应该是自上而下，我们这里写的比较简单所以看不出来，我们来稍微复杂一点。

In [13]:
# 定义函数
def devide(a,b):
    return a / b

print(devide(1,0))

ZeroDivisionError: division by zero

# 解读

这样我们是不是可以清晰的看到先是第5行`print()`里面出错了，再深究下去，就是定义的`devide`函数出错了，什么错误？是「除零错误」。同时你也可以看到`in <module>`和`in devide(a,b)`都是明确表明错误发生的位置的。

module 翻译为模组，跟它的发音很像，表示错误发生在代码的最外层。

当然我们这里的问题比较简单，比较复杂的时候就需要我们根据这些提示来好好分析了。

关于函数定义 def，我们在 Python 进阶里面的[定义函数]()会讲到。

# 其他异常

除了常见的，还有很多异常类型，

未来在你自己编写的过程中也会遇到各种各样的异常。

但是通常这些异常都可以通过以上方式来解决。

# 异常解决的好方法

遇到异常，最简单直接的方式就是把异常拿到搜索引擎，找到最佳的解决方案，通常最好的答案会在这里：
[stackoverflow](https://stackoverflow.com/)

当然你也可以研究：
[python 官方文档](https://docs.python.org/3/)

# 异常的预判和防护

让我来回到我们之前提到的一个小游戏，就是下面这段代码：

In [17]:
import random

num = random.randint(1,100)  # 获取一个1-100间的随机数
is_done = False # 是否猜中的标记
count = 0 # 玩家玩了几次

while not is_done:
    guess = int(input('请输入一个[1,100]的整数：'))
    if guess == num:
        is_done = True
    elif guess > num:
        print('大了，再猜猜')
    elif guess < num:
        print('小了，再猜猜')
    count += 1
    
print('恭喜你，猜了{}次，终于猜对了！答案是{}!'.format(count,num))

请输入一个[1,100]的整数：10000
大了，再猜猜
请输入一个[1,100]的整数：1.23


ValueError: invalid literal for int() with base 10: '1.23'

# 解读

第 8 行，我们用 input 函数来接受数据并返回字符串，用 int 函数讲字符串转换成[整形]()

那要是有个用户调皮，他没有输入整数，他输入了其他的字符串呢？

（**动手试试吧！**）

试过之后，我们可以看到，程序抛出了一个 `ValueError`的错误，具体解释是「1.2 这个无效输入，无法被转换成 10 进制的整数」。

因为我们没有对异常进行任何处理，所以当用户